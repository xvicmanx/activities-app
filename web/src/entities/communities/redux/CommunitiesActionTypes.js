export const SET_COMMUNITIES = 'SET_COMMUNITIES';
export const SET_COMMUNITIES_LOADING_STATE = 'SET_COMMUNITIES_LOADING_STATE';
export const SET_COMMUNITIES_ERROR = 'SET_COMMUNITIES_ERROR';

export const SET_COMMUNITY = 'SET_COMMUNITY';
export const SET_COMMUNITY_LOADING_STATE = 'SET_COMMUNITY_LOADING_STATE';
export const SET_COMMUNITY_ERROR = 'SET_COMMUNITY_ERROR';

export const ADD_MEMBER = 'ADD_MEMBER';
