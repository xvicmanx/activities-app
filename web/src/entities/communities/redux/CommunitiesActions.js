import {
  SET_COMMUNITIES,
  SET_COMMUNITIES_LOADING_STATE,
  SET_COMMUNITIES_ERROR,
  SET_COMMUNITY,
  SET_COMMUNITY_LOADING_STATE,
  SET_COMMUNITY_ERROR,
  ADD_MEMBER,
} from './CommunitiesActionTypes';

import CommunitiesService from '../services/CommunitiesService';

export const fetchCommunities = (token) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITIES_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.fetchCommunities(token);

  if (res.success) {
    dispatch({
      type: SET_COMMUNITIES,
      payload: res.communities,
    });
  } else {
    dispatch({
      type: SET_COMMUNITIES_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_COMMUNITIES_LOADING_STATE,
    payload: false,
  });
};


export const fetchCommunity = (token, id) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.fetchCommunity(token, id);

  if (res.success) {
    dispatch({
      type: SET_COMMUNITY,
      payload: res.community,
    });
  } else {
    dispatch({
      type: SET_COMMUNITY_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: false,
  });
};


export const addMember = (token, communityId, memberId, coordinates) => async (dispatch) => {
  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: true,
  });

  const res = await CommunitiesService.addMember(token, communityId, memberId, coordinates);

  if (res && res.success) {
    dispatch({
      type: ADD_MEMBER,
      payload: res.member,
    });
  } else {
    dispatch({
      type: SET_COMMUNITY_ERROR,
      payload: res ? res.message : 'Unexpected error',
    });
  }

  dispatch({
    type: SET_COMMUNITY_LOADING_STATE,
    payload: false,
  });
};