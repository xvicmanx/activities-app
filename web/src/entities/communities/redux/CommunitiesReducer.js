import {
  SET_COMMUNITIES,
  SET_COMMUNITIES_LOADING_STATE,
  SET_COMMUNITIES_ERROR,
  SET_COMMUNITY,
  SET_COMMUNITY_LOADING_STATE,
  SET_COMMUNITY_ERROR,
  ADD_MEMBER,
} from './CommunitiesActionTypes';

const INITIAL_STATE = {
  communities: null,
  community: null,
  member: null,
  loading: false,
  error: null,
};

const CommunitiesReducer = (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case SET_COMMUNITIES_LOADING_STATE:
      return {
        ...state,
        loading: payload,
        error: payload ? null : state.error,
      };

    case SET_COMMUNITIES_ERROR:
      return {
        ...state,
        error: payload,
      };

    case SET_COMMUNITIES:
      return {
        ...state,
        communities: payload,
      };

    case ADD_MEMBER:
      return {
        ...state,
        community: {
          ...state.community,
          members: [...(state.community.members || []), payload],
        },
      };

    case SET_COMMUNITY_LOADING_STATE:
      return {
        ...state,
        loading: payload,
        error: payload ? null : state.error,
      };

    case SET_COMMUNITY_ERROR:
      return {
        ...state,
        error: payload,
      };

    case SET_COMMUNITY:
      return {
        ...state,
        community: payload,
      };

    default:
      return state;
  }
};

export default CommunitiesReducer;
