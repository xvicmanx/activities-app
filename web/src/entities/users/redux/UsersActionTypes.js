export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const LOG_OUT_CURRENT_USER = 'LOG_OUT_CURRENT_USER';
export const SET_USERS_LOADING_STATE = 'SET_USERS_LOADING_STATE';
export const SET_USER_ERROR = 'SET_USER_ERROR';
export const SET_USERS = 'SET_USERS';
export const SET_USERS_ERROR = 'SET_USERS_ERROR';
