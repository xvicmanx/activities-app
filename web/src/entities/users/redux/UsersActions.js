import cookies from 'browser-cookies';
import {
  SET_CURRENT_USER,
  LOG_OUT_CURRENT_USER,
  SET_USERS_LOADING_STATE,
  SET_USER_ERROR,
  SET_USERS_ERROR,
  SET_USERS,
} from './UsersActionTypes';

import UsersService from '../services/UsersService';

const storeToken = (token) => {
  cookies.set('jwt', token, { expires: 1 });
};

export const clearToken = () => {
  cookies.erase('jwt');
};

export const readToken = () => cookies.get('jwt');

export const loadSignedInUser = ({ shouldLoad }) => async (dispatch) => {
  if (!shouldLoad) {
    return;
  }

  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: true,
  });

  const res = await UsersService.loadSignedInUser();

  clearToken();

  if (res.success) {
    storeToken(res.token);
    dispatch({
      type: SET_CURRENT_USER,
      payload: { ...res.user, token: res.token },
    });
  } else {
    dispatch({
      type: SET_USER_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: false,
  });
};

export const loginUser = (email, password) => async (dispatch) => {
  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: true,
  });

  const res = await UsersService.loginUser(email, password);

  if (res.success) {
    storeToken(res.token);
    dispatch({
      type: SET_CURRENT_USER,
      payload: { ...res.user, token: res.token },
    });
  } else {
    dispatch({
      type: SET_USER_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: false,
  });
};

export const logOutUser = () => async (dispatch) => {
  clearToken();
  dispatch({
    type: LOG_OUT_CURRENT_USER,
  });
};

export const fetchUsers = () => async (dispatch) => {
  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: true,
  });

  const res = await UsersService.fetchUsers();

  if (res.success) {
    dispatch({
      type: SET_USERS,
      payload: res.users,
    });
  } else {
    dispatch({
      type: SET_USERS_ERROR,
      payload: res.message,
    });
  }

  dispatch({
    type: SET_USERS_LOADING_STATE,
    payload: false,
  });
};
