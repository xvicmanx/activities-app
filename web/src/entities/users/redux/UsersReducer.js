import {
  SET_CURRENT_USER,
  LOG_OUT_CURRENT_USER,
  SET_USERS_LOADING_STATE,
  SET_USER_ERROR,
  SET_USERS_ERROR,
  SET_USERS,
} from './UsersActionTypes';

export const INITIAL_STATE = {
  data: null,
  users: null,
  loading: false,
  error: null,
};

const UsersReducer = (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case SET_USERS_LOADING_STATE:
      return {
        ...state,
        loading: payload,
        error: payload ? null : state.error,
      };

    case SET_USER_ERROR:
    case SET_USERS_ERROR:
      return {
        ...state,
        error: payload,
      };

    case SET_CURRENT_USER:
      return {
        ...state,
        data: payload,
      };

    case SET_USERS:
      return {
        ...state,
        users: payload,
      };

    case LOG_OUT_CURRENT_USER:
      return {
        ...state,
        data: null,
      };

    default:
      return state;
  }
};

export default UsersReducer;
