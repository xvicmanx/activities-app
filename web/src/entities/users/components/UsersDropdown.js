// @flow

import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { fetchUsers, readTokenFromCookie } from '../redux/UsersActions';

import { DropDown, LoadingIndicator } from '../../../core/components';

type Props = {
  name: string,
  value?: string | number,
  onChange: Function,
};

const UsersDropdown = ({
  name,
  value,
  onChange,
}: Props): React$Element<any> => {
  const { Users } = useSelector((state) => state);
  const dispatch = useDispatch();
  const token = readTokenFromCookie();

  useEffect(() => {
    dispatch(fetchUsers(token));
  }, [dispatch, token]);

  if (Users.isLoading) {
    return <LoadingIndicator />;
  }

  return (
    <DropDown
      name={name}
      items={Users.users || []}
      onChange={onChange}
      placeholder="Seleccione usuario"
      value={value}
    />
  );
};

export default UsersDropdown;
