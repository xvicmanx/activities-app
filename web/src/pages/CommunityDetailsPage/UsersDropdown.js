import React, { useEffect } from 'react';
import { Heading } from 'react-bulma-components';
import { useSelector, useDispatch } from 'react-redux';

import { fetchUsers } from '../../redux/Users/UsersActions';

import {
  DropDown,
  LoadingIndicator,
} from '../../components';
import { readTokenFromCookie } from '../../redux/Users/UsersActions';

const UsersDropdown = ({ name, value, onChange }) => {
  const { Users } = useSelector(state => state);
  const dispatch = useDispatch();
  const token = readTokenFromCookie();

  useEffect(() => {
    dispatch(fetchUsers(token));
  }, [dispatch]);

  if (Users.isLoading) {
    return <LoadingIndicator />;
  }

  return (
    <DropDown
      name={name}
      items={Users.users || []}
      onChange={onChange}
      placeholder="Seleccione usuario"
      value={value}
    />
  );
};

export default UsersDropdown;
