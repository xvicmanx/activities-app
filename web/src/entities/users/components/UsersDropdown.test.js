import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

Enzyme.configure({ adapter: new Adapter() });

import UsersDropdown from './UsersDropdown';

jest.mock('../redux/UsersActions');

import { fetchUsers, readToken } from '../redux/UsersActions';


const mockStore = configureMockStore();

describe('UsersDropdown', () => {
  const token = 'test-token';
  const users = [
    {
      id: 1,
      name: 'John',
    },
    {
      id: 2,
      name: 'Mary',
    },
  ];

  beforeEach(() => {
    readToken.mockClear().mockImplementation(() => token);
    fetchUsers.mockClear().mockImplementation(() => ({
      type: 'SET_USERS',
      payload: users,
    }));
  });

  it('renders properly', () => {
    const store = mockStore({
      Users: { users },
    });
    const onChange = jest.fn();
    const result = Enzyme.mount(
      <Provider store={store}>
        <UsersDropdown
          name="test"
          value={1}
          onChange={onChange}
        />
      </Provider>
    );

    expect(readToken).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledWith(token);
    expect(result.find('DropDown').props()).toMatchSnapshot();
  });

  it('renders properly even if the users list is not present', () => {
    const store = mockStore({
      Users: {},
    });
    const onChange = jest.fn();
    const result = Enzyme.mount(
      <Provider store={store}>
        <UsersDropdown
          name="test"
          value={1}
          onChange={onChange}
        />
      </Provider>
    );

    expect(readToken).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledWith(token);
    expect(result.find('DropDown').props()).toMatchSnapshot();
  });

  it('renders properly when loading data', () => {
    const store = mockStore({
      Users: { isLoading: true },
    });
    const onChange = jest.fn();
    const result = Enzyme.mount(
      <Provider store={store}>
        <UsersDropdown
          name="test"
          value={1}
          onChange={onChange}
        />
      </Provider>
    );

    expect(readToken).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledTimes(1);
    expect(fetchUsers).toHaveBeenCalledWith(token);
    expect(result.find('DropDown').length).toBe(0);
    expect(result.find('LoadingIndicator').length).toBe(1);
  });
});
