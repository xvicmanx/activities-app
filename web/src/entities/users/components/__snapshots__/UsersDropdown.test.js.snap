// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersDropdown renders properly 1`] = `
Object {
  "items": Array [
    Object {
      "id": 1,
      "name": "John",
    },
    Object {
      "id": 2,
      "name": "Mary",
    },
  ],
  "name": "test",
  "onChange": [MockFunction],
  "placeholder": "Seleccione usuario",
  "value": 1,
}
`;

exports[`UsersDropdown renders properly even if the users list is not present 1`] = `
Object {
  "items": Array [],
  "name": "test",
  "onChange": [MockFunction],
  "placeholder": "Seleccione usuario",
  "value": 1,
}
`;
